Make C API docs publishable

frankharkins opened the pull request:
### Summary

This PR adds the config needed to build the C API docs, and makes some tweaks to the docs themselves to improve appearance on the documentation website.

frankharkins committed:
Get C API docs building

Co-authored-by: Max Rossmannek <rmax@ethz.ch>
Co-authored-by: Julien Gacon <jul@zurich.ibm.com>

frankharkins committed:
Separate `docs` from `docs-c` in `tox.ini`

frankharkins committed:
Move each group to its own page

frankharkins committed:
Reorganise docs page layout

frankharkins committed:
Fix some group namings

frankharkins committed:
Get enums working

frankharkins committed:
Update tox docs-clean

frankharkins committed:
Add exit codes page

frankharkins committed:
Add "Data types" heading

mtreinish commented on docs/cdoc/index.rst:
```diff
@@ -0,0 +1,14 @@
+Qiskit C API (``qiskit.h``)
+===========================
+
+Quantum information:
```
I think we should add some more details to this page because it's basically the home page for the C API:

```suggestion
The Qiskit C API is a low level interface to the core data model of Qiskit. It is designed to provide
a high performance interface to Qiskit for compiled languages and provides a defined ABI to the
internal Rust data model that is used to create the Python API. There are two expected modes of
operation for the C API:

 * A standalone shared library for creating and working with Qiskit objects from compiled
    languages without a runtime dependency on Python, and
 * For building Python extensions that are using Qiskit but interface directly with the Rust
    objects from the extension code without using Python for better performance.
 
More details on building/installing the C API can be found at:

https://docs.quantum.ibm.com/guides/install-c-api

and for combining the C API with custom Python extensions can be found at:

https://docs.quantum.ibm.com/guides/c-extension-for-python

As this interface is still new in Qiskit it should be considered experimental and the interface might change between minor version releases.

Quantum information
------------------
```

mtreinish commented on docs/cdoc/index.rst:
```diff
@@ -0,0 +1,14 @@
+   qk-obs
+   qk-obs-term
+   qk-bit-term
+   exit-codes
```
We might want to split this out into a separate toctree because it will be used by more than the sparse observable (see #14006). But we can do that later when we update the docs in a follow up.

mtreinish reviewed (commented):
Overall I think this is looking good, I left a couple small inline comments about the content but I think the build infrastructure looks good

frankharkins committed:
Update docs/cdoc/index.rst

Co-authored-by: Matthew Treinish <mtreinish@kortar.org>

Cryoris commented:
We need to find a way to add custom docs to opaque objects, like ``QkObs``. Right now this appears in the header as typedef with the Rust documentation. Some options could be to 
1. add custom docs in the `@defgroup` definition (in `docs/cdoc/index.h`) and remove the `typedef` documentation
2. add a custom typedef in the `cext` crate just for the sake of docs, e.g. something like
    ```rust
    use qiskit_accelerate::sparse_observable::SparseObservable;

    /// @ingroup QkObs
    /// Custom C docstring for QkObs goes here....
    type CSparseObservable = SparseObservable;

    // use CSparseObservable throughout
    ```
    Note that I didn't test this solution to see what pops up in the header yet
3. Somehow update the header post-building to inject the correct docstring 

I'm not super happy with either of these... 1 could be a good compromise, if it is possible to skip ``QkObs`` typedef in the docs (is it @frankharkins? 🙂). 2 is not amazing since we'd have to add a typedef just for docs, but it would be fine too I think. 3 sounds very hacky.

frankharkins commented:
You can remove the typedef by removing the `doxygentypedef`:

https://github.com/Qiskit/qiskit/blob/55610d2ca5aef23a4206b81b43cc6644b192a8f9/docs/cdoc/qk-obs.rst?plain=1#L12

Then you could just write whatever you like there, although it won't be formatted as an object in the docs (that is, with the blue border and signature).

<img width="630" alt="Screenshot 2025-03-27 at 14 01 54" src="https://github.com/user-attachments/assets/c6d5ba68-b6e1-436f-af8a-9b29c49fcfa1" />

mtreinish commented:
It looks like sphinx as run in CI isn't quite happy yet:

https://dev.azure.com/qiskit-ci/qiskit/_build/results?buildId=66017&view=logs&j=ee6a574e-a884-556d-e11f-f4e0408370dd&t=cda6e954-50e0-5d66-9385-47256a656c01&l=258

The warnings listed there are all treated as fatal in how we have sphinx configured. I assume this is because we're only running  `tox -edocs` and we're not calling doxygen. If the intent is to have the C docs as an optional extra we might need a bit more infrastructure to exclude it from the base sphinx build or otherwise maybe just remove the job split between docs and c-docs.

Cryoris committed:
QkObs docs and nits

Cryoris committed:
add bit term

note that the Members section should be updated to match doxygen's enum style before merging, the table is a placeholder

frankharkins committed:
Re-format index RST

Cryoris committed:
update members

frankharkins committed:
Merge pull request #1 from Cryoris/c-api-docs-frank

QkObs docs and nits

Cryoris committed:
iterating on layout

* better enums (follow Glib style because it looks great)
* fix linebreaks
* minor fixes (brackets, missing whitespace)

Cryoris committed:
try fix tables

frankharkins committed:
Fix RST formatting quirks

frankharkins committed:
Install doxygen in CI

Cryoris committed:
document exit codes

Cryoris committed:
Merge commit 'refs/pull/14101/head' of github.com:Qiskit/qiskit into c-api-docs-frank

Cryoris committed:
rm trailing ]

frankharkins committed:
Merge pull request #2 from Cryoris/c-api-docs-frank

Update enums and fix tables

Cryoris committed:
more enum work

frankharkins committed:
Fix table

Cryoris committed:
QkExitCode and remove rustdoc links

frankharkins committed:
Merge branch 'FH/c-api-docs' of https://github.com/frankharkins/qiskit-terra into FH/c-api-docs

frankharkins committed:
Rename exit codes page

Cryoris commented on docs/cdoc/index.rst:
```diff
@@ -0,0 +1,35 @@
+   qk-obs
+   qk-obs-term
+   qk-bit-term
+   exit-codes
```
Should this be `qk-exit-codes`? CI is complaining about not finding `exit-codes` and `qk-exit-codes` not being in a toctree
```suggestion
   qk-exit-codes
```

frankharkins committed:
Update docs/cdoc/index.rst

Co-authored-by: Julien Gacon <gaconju@gmail.com>

frankharkins committed:
Fix xrefs and toc

frankharkins committed:
Merge branch 'FH/c-api-docs' of https://github.com/frankharkins/qiskit-terra into FH/c-api-docs

frankharkins committed:
Fix CI?

frankharkins committed:
Remove bad xrefs

Cryoris commented on docs/cdoc/qk-obs.rst:
```diff
@@ -0,0 +1,253 @@
+            qk_obs_free(diff);
+            qk_obs_free(canonical);
+            qk_obs_free(zero);
+            return equal;
```
Tiny typo, otherwise LGTM!
```suggestion
            return equiv;
```

Cryoris committed:
Update docs/cdoc/qk-obs.rst

mtreinish committed:
Document doxygen is required for docs builds now

mtreinish reviewed (approved):
This LGTM now, I pushed up a small commit to the CONTRIBUTING.md file to document doxygen is needed now. But I don't think there is anything else needed here. We can continue to iterate on the content in follow up PRs if there is more needed there or other changes.
